File: androidApp/src/main/java/com/cocktailcraft/android/viewmodel/OfflineModeViewModelSKIE.kt
package com.cocktailcraft.android.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.cocktailcraft.domain.model.Cocktail
import com.cocktailcraft.viewmodel.SharedOfflineModeViewModel
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

/**
 * Android ViewModel wrapper for SharedOfflineModeViewModel using SKIE.
 * This replaces the old OfflineModeViewModel with a thin wrapper around the shared implementation.
 * 
 * Key SKIE benefits:
 * - StateFlows work directly without conversion
 * - Suspend functions work with Android coroutines
 * - Shared business logic reduces code duplication
 * - Type safety preserved across platforms
 */
class OfflineModeViewModelSKIE : ViewModel(), KoinComponent {
    
    // Inject the shared ViewModel
    private val sharedViewModel: SharedOfflineModeViewModel by inject()
    
    // Expose StateFlows from shared ViewModel (SKIE handles the interop)
    val isOfflineModeEnabled: StateFlow<Boolean> = sharedViewModel.isOfflineModeEnabled
    val isNetworkAvailable: StateFlow<Boolean> = sharedViewModel.isNetworkAvailable
    val recentlyViewedCocktails: StateFlow<List<Cocktail>> = sharedViewModel.recentlyViewedCocktails
    val cacheSize: StateFlow<Int> = sharedViewModel.cacheSize
    val lastSyncTime: StateFlow<String?> = sharedViewModel.lastSyncTime
    
    // Expose loading and error from shared base class
    val loadingState: StateFlow<Boolean> = sharedViewModel.isLoading
    val errorState = sharedViewModel.error
    val errorEventState = sharedViewModel.errorEvent
    
    // Computed properties
    val hasRecentlyViewed: Boolean
        get() = sharedViewModel.hasRecentlyViewed
    
    val isOnlineAndOfflineModeDisabled: Boolean
        get() = sharedViewModel.isOnlineAndOfflineModeDisabled
    
    // MARK: - Public Methods (SKIE converts these to proper suspend functions)
    
    /**
     * Toggle offline mode using SKIE async interop
     */
    fun toggleOfflineMode() {
        viewModelScope.launch {
            sharedViewModel.toggleOfflineMode()
        }
    }
    
    /**
     * Set offline mode using SKIE async interop
     */
    fun setOfflineMode(enabled: Boolean) {
        viewModelScope.launch {
            sharedViewModel.setOfflineMode(enabled)
        }
    }
    
    /**
     * Sync cached data with server using SKIE async interop
     */
    fun syncCachedData() {
        viewModelScope.launch {
            sharedViewModel.syncCachedData()
        }
    }
    
    /**
     * Clear all cached data using SKIE async interop
     */
    fun clearCache() {
        viewModelScope.launch {
            sharedViewModel.clearCache()
        }
    }
    
    /**
     * Load recently viewed cocktails using SKIE async interop
     */
    fun loadRecentlyViewedCocktails() {
        viewModelScope.launch {
            sharedViewModel.loadRecentlyViewedCocktails()
        }
    }
    
    // MARK: - Synchronous Methods (direct delegation)
    
    /**
     * Get cached cocktail count
     */
    fun getCachedCocktailCount(): Int {
        return sharedViewModel.getCachedCocktailCount()
    }
    
    /**
     * Get network status as string
     */
    fun getNetworkStatus(): String {
        return sharedViewModel.getNetworkStatus()
    }
    
    /**
     * Get offline mode status as string
     */
    fun getOfflineModeStatus(): String {
        return sharedViewModel.getOfflineModeStatus()
    }
    
    /**
     * Get recently viewed cocktails by category
     */
    fun getRecentlyViewedByCategory(category: String): List<Cocktail> {
        return sharedViewModel.getRecentlyViewedByCategory(category)
    }
    
    /**
     * Get recently viewed cocktails with limit
     */
    fun getRecentlyViewedWithLimit(limit: Int): List<Cocktail> {
        return sharedViewModel.getRecentlyViewedWithLimit(limit)
    }
    
    /**
     * Check if offline mode is recommended
     */
    fun isOfflineModeRecommended(): Boolean {
        return sharedViewModel.isOfflineModeRecommended()
    }
    
    /**
     * Get cache status summary
     */
    fun getCacheStatusSummary(): String {
        return sharedViewModel.getCacheStatusSummary()
    }
    
    /**
     * Clear the current error
     */
    fun clearErrorState() {
        sharedViewModel.clearError()
    }
    
    /**
     * Refresh offline mode data
     */
    fun refresh() {
        sharedViewModel.refresh()
    }
    
    /**
     * Clean up when ViewModel is cleared
     */
    override fun onCleared() {
        super.onCleared()
        sharedViewModel.onCleared()
    }
}

File: iosApp/CocktailCraft/ViewModels/OfflineModeViewModelSKIE.swift
import SwiftUI
import shared

/**
 * iOS ViewModel wrapper for SharedOfflineModeViewModel using pure SKIE integration.
 * State/error mirroring and observation-task lifecycle live in
 * SharedViewModelWrapper.
 */
final class OfflineModeViewModelSKIE: SharedViewModelWrapper<OfflineUiState> {

    // Computed properties
    var hasRecentlyViewed: Bool {
        !state.recentlyViewedCocktails.isEmpty
    }

    var isOnlineAndOfflineModeDisabled: Bool {
        state.isNetworkAvailable && !state.isOfflineModeEnabled
    }

    var cacheStatusSummary: String {
        sharedViewModel.getCacheStatusSummary()
    }

    var networkStatus: String {
        sharedViewModel.getNetworkStatus()
    }

    var offlineModeStatus: String {
        sharedViewModel.getOfflineModeStatus()
    }

    // Shared ViewModel instance
    private let sharedViewModel: SharedOfflineModeViewModel

    init() {
        let viewModel = getSharedKoinHelper().getSharedOfflineModeViewModel()
        self.sharedViewModel = viewModel
        super.init(uiState: viewModel.uiState, errorFlow: viewModel.error)
    }

    // No deinit: the base class cancels observation. Wraps a Koin `single`
    // whose coroutine scope must survive any one wrapper — never onCleared().

    // MARK: - Public Methods (using SKIE async/await)

    func toggleOfflineMode() async {
        do {
            try await sharedViewModel.toggleOfflineMode()
        } catch {
            print("OfflineModeViewModelSKIE - Error toggling offline mode: \(error)")
        }
    }

    func setOfflineMode(_ enabled: Bool) async {
        do {
            try await sharedViewModel.setOfflineMode(enabled: enabled)
        } catch {
            print("OfflineModeViewModelSKIE - Error setting offline mode: \(error)")
        }
    }

    func syncCachedData() async {
        do {
            try await sharedViewModel.syncCachedData()
        } catch {
            print("OfflineModeViewModelSKIE - Error syncing cached data: \(error)")
        }
    }

    func clearCache() async {
        do {
            try await sharedViewModel.clearCache()
        } catch {
            print("OfflineModeViewModelSKIE - Error clearing cache: \(error)")
        }
    }

    func loadRecentlyViewedCocktails() async {
        do {
            try await sharedViewModel.loadRecentlyViewedCocktails()
        } catch {
            print("OfflineModeViewModelSKIE - Error loading recently viewed cocktails: \(error)")
        }
    }

    // MARK: - Synchronous Methods

    func getCachedCocktailCount() -> Int {
        return Int(sharedViewModel.getCachedCocktailCount())
    }

    func getRecentlyViewedByCategory(_ category: String) -> [Cocktail] {
        return sharedViewModel.getRecentlyViewedByCategory(category: category)
    }

    func getRecentlyViewedWithLimit(_ limit: Int) -> [Cocktail] {
        return sharedViewModel.getRecentlyViewedWithLimit(limit: Int32(limit))
    }

    func isOfflineModeRecommended() -> Bool {
        return sharedViewModel.isOfflineModeRecommended()
    }

    func clearError() {
        sharedViewModel.clearError()
    }

    func refresh() {
        sharedViewModel.refresh()
    }

    // MARK: - Helper Methods for SwiftUI

    func formatLastSyncTime() -> String {
        guard let syncTime = state.lastSyncTime else {
            return "Never synced"
        }
        return "Last sync: \(syncTime)"
    }

    func getCacheSizeText() -> String {
        let count = getCachedCocktailCount()
        return "\(count) cocktail\(count == 1 ? "" : "s") cached"
    }

    func getNetworkStatusColor() -> Color {
        return state.isNetworkAvailable ? .green : .red
    }

    func getOfflineModeColor() -> Color {
        return state.isOfflineModeEnabled ? .orange : .blue
    }

    func shouldShowOfflineRecommendation() -> Bool {
        return !state.isNetworkAvailable && !state.isOfflineModeEnabled
    }
}
